### PR TITLE
Feature/mattermost-integration

### PR DESCRIPTION
**Pull request by Itaykal**
Add Mattermost integration

### CODE REVIEW

**Comment by yuvalsteuer** [comment]
Prove it works.
בלי תמונה אין אמונה

**Comment by Itaykal** [comment]
image

File: app/data_sources/mattermost.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+            for id, post in posts["posts"].items():
+                if not self._is_valid_message(post):
+                    continue
+                
+                author = self._mattermost.users.get_user(post["user_id"])["username"]
+                author_image_url = f"{self._get_mattermost_url()}/api/v4/users/{post['user_id']}/image?_=0"
+                content = post["message"]
+                timestamp = datetime.fromtimestamp(post["update_at"] / 1000)
+                parsed_posts.append(
```
Comment (Roey7) [inline app/data_sources/mattermost.py:135]: If a user sent 4 messages in a row we would like to consider them a single BasicDocumnt.
You can look at Slack implementation as a reference

File: app/data_sources/mattermost.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+    
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        mattermost_config = MattermostConfig(**self._config)
+        self._mattermost = Driver(options = asdict(mattermost_config))
+    
+    
+    def _list_channels(self) -> List[MattermostChannel]:
+        channels = self._mattermost.channels.client.get("/channels")
```
Comment (Roey7) [inline app/data_sources/mattermost.py:72]: do we need pagination here?

File: app/data_sources/mattermost.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        
+        team_url = self._get_team_url(channel)
+        while True:
+            posts = self._list_posts_in_channel(channel.id, page)
+
+            for id, post in posts["posts"].items():
+                if not self._is_valid_message(post):
+                    continue
+                
+                author = self._mattermost.users.get_user(post["user_id"])["username"]
```
Comment (Roey7) [inline app/data_sources/mattermost.py:131]: Cache mechanism is required here 
look at google drive, Bary used a nice builtin decorator for that

File: app/data_sources/mattermost.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+        
+
+class MattermostDataSource(BaseDataSource):
+    FEED_BATCH_SIZE = 500
+
+    @staticmethod
+    def get_config_fields() -> List[ConfigField]:
+        return [
+            ConfigField(label="Mattermost Server", name="url", placeholder="https://mattermost.server.com", input_type=HTMLInputType.TEXT),
```
Comment (Roey7) [inline app/data_sources/mattermost.py:50]: Please provide instructions to add mattermost, you should add them to data-source-panel.tsx

File: app/data_sources/mattermost.py (hunk 2 above)
Comment (Itaykal) [reply app/data_sources/mattermost.py:72]: Defaults to getting all channels, not sure if it's the smartest move

File: app/data_sources/mattermost.py (hunk 3 above)
Comment (Itaykal) [reply app/data_sources/mattermost.py]: Done

File: app/data_sources/mattermost.py (hunk 1 above)
Comment (Itaykal) [reply app/data_sources/mattermost.py]: Done

File: app/data_sources/mattermost.py (hunk 4 above)
Comment (Itaykal) [reply app/data_sources/mattermost.py:51]: Done

**Pull request merged**